File: internal/voip/media/videoframe.go
package media

import "encoding/binary"

// VideoFrame é um quadro H264 (Annex-B) codificado, trocado com o browser pelo
// data channel "vp8". O browser codifica/decodifica com WebCodecs; o lado Go só
// empacota em RTP, então o bitstream do codec é opaco aqui.
type VideoFrame struct {
	Keyframe    bool
	TimestampMS uint32
	// Rotation é a orientação do quadro em graus no sentido horário
	// (0, 90, 180 ou 270), vinda da extensão RTP CVO do WhatsApp. O browser
	// gira o desenho no canvas de acordo.
	Rotation uint16
	Data     []byte
}

// videoFrameHeaderLen é o prefixo fixo de toda mensagem do data channel "vp8":
// 1 byte de flags + 4 bytes de timestamp em ms (big-endian).
//
// Flags: bit 0 = keyframe; bits 1-2 = código de rotação (0=0°, 1=90°, 2=180°,
// 3=270°).
const videoFrameHeaderLen = 5

func rotationToCode(deg uint16) byte {
	switch deg {
	case 90:
		return 1
	case 180:
		return 2
	case 270:
		return 3
	default:
		return 0
	}
}

func codeToRotation(code byte) uint16 {
	switch code & 0x03 {
	case 1:
		return 90
	case 2:
		return 180
	case 3:
		return 270
	default:
		return 0
	}
}

// EncodeVideoFrame serializa um quadro para o data channel.
func EncodeVideoFrame(f VideoFrame) []byte {
	out := make([]byte, videoFrameHeaderLen+len(f.Data))
	if f.Keyframe {
		out[0] |= 1
	}
	out[0] |= rotationToCode(f.Rotation) << 1
	binary.BigEndian.PutUint32(out[1:], f.TimestampMS)
	copy(out[videoFrameHeaderLen:], f.Data)
	return out
}

// videoCtlKeyframeRequest é o corpo de uma mensagem de controle no data channel
// "vp8" (server → browser) pedindo um keyframe imediato — disparada por um
// RTCP PLI/FIR do peer. Toda mensagem de quadro tem >= videoFrameHeaderLen (5)
// bytes, então uma mensagem de exatamente 1 byte nunca é confundida com quadro.
const videoCtlKeyframeRequest = 0x01

// VideoKeyframeRequestMsg devolve o corpo da mensagem de controle "manda um
// keyframe agora" para o data channel "vp8".
func VideoKeyframeRequestMsg() []byte { return []byte{videoCtlKeyframeRequest} }

// IsVideoKeyframeRequest reconhece a mensagem de controle de pedido de keyframe.
func IsVideoKeyframeRequest(b []byte) bool {
	return len(b) == 1 && b[0] == videoCtlKeyframeRequest
}

// DecodeVideoFrame faz o parse de uma mensagem do data channel. ok é false
// quando a mensagem é curta demais para o cabeçalho.
func DecodeVideoFrame(b []byte) (f VideoFrame, ok bool) {
	if len(b) < videoFrameHeaderLen {
		return VideoFrame{}, false
	}
	f.Keyframe = b[0]&1 != 0
	f.Rotation = codeToRotation(b[0] >> 1)
	f.TimestampMS = binary.BigEndian.Uint32(b[1:])
	f.Data = append([]byte(nil), b[videoFrameHeaderLen:]...)
	return f, true
}

File: client/src/lib/video-frame.ts
// Formato de fio do data channel "vp8", espelho byte a byte de
// media.EncodeVideoFrame / media.DecodeVideoFrame em
// internal/voip/media/videoframe.go: 1 byte de flags + 4 bytes de timestamp em
// ms (big-endian) + o bitstream H264 (Annex-B) opaco.
//
// Flags: bit 0 = keyframe; bits 1-2 = rotação (0=0°, 1=90°, 2=180°, 3=270°,
// sentido horário).

export type WireVideoFrame = {
  keyframe: boolean;
  rotationDeg: 0 | 90 | 180 | 270;
  timestampMs: number;
  data: Uint8Array;
};

const HEADER_LEN = 5;
const ROT_BY_CODE: Record<number, 0 | 90 | 180 | 270> = { 0: 0, 1: 90, 2: 180, 3: 270 };
const CODE_BY_ROT: Record<number, number> = { 0: 0, 90: 1, 180: 2, 270: 3 };

export const encodeVideoFrame = (f: WireVideoFrame): ArrayBuffer => {
  const out = new Uint8Array(HEADER_LEN + f.data.byteLength);
  const view = new DataView(out.buffer);
  out[0] = (f.keyframe ? 1 : 0) | ((CODE_BY_ROT[f.rotationDeg] ?? 0) << 1);
  view.setUint32(1, f.timestampMs >>> 0, false);
  out.set(f.data, HEADER_LEN);
  return out.buffer;
};

// decodeVideoFrame faz o parse de uma mensagem do data channel. Retorna null
// quando a mensagem é curta demais para o cabeçalho (equivale a ok=false no Go).
export const decodeVideoFrame = (buf: ArrayBuffer): WireVideoFrame | null => {
  if (buf.byteLength < HEADER_LEN) return null;
  const view = new DataView(buf);
  const flags = view.getUint8(0);
  return {
    keyframe: (flags & 1) !== 0,
    rotationDeg: ROT_BY_CODE[(flags >> 1) & 3] ?? 0,
    timestampMs: view.getUint32(1, false),
    data: new Uint8Array(buf, HEADER_LEN),
  };
};
